crates/cxx-qt-lib/src/core/qtlogging.rs

```diff
@@ -0,0 +1,138 @@
+        QtCriticalMsg = 2,
+        QtSystemMsg = 2
+    }
+
+    unsafe extern "C++" {
+        include!("cxx-qt-lib/common.h");
+
+        include!("cxx-qt-lib/qstring.h");
+        type QString = crate::QString;
+
+        include!("cxx-qt-lib/qtlogging.h");
+        type QMessageLogContext = crate::QMessageLogContext;
+        type QtMsgType;
+
+        fn qt_message_output(msgType: QtMsgType, context: &QMessageLogContext, string: &QString);
+
+        #[cxx_name = "qmessagelogcontext_line"]
+        fn line(context: &QMessageLogContext) -> i32;
+
+        #[cxx_name = "qmessagelogcontext_set_line"]
+        fn set_line(context: &mut QMessageLogContext, line: i32);
+
+        #[cxx_name = "qmessagelogcontext_file"]
+        unsafe fn file(context: &QMessageLogContext) -> *const c_char;
+
+        #[cxx_name = "qmessagelogcontext_set_file"]
+        unsafe fn set_file(context: &mut QMessageLogContext, file: *const c_char);
+
+        #[cxx_name = "qmessagelogcontext_function"]
+        unsafe fn function(context: &QMessageLogContext) -> *const c_char;
+
+        #[cxx_name = "qmessagelogcontext_set_function"]
+        unsafe fn set_function(context: &mut QMessageLogContext, function: *const c_char);
+
+        #[cxx_name = "qmessagelogcontext_category"]
+        unsafe fn category(context: &QMessageLogContext) -> *const c_char;
+
+        #[cxx_name = "qmessagelogcontext_set_category"]
+        unsafe fn set_category(context: &mut QMessageLogContext, category: *const c_char);
+    }
+
+    #[namespace = "rust::cxxqtlib1"]
+    unsafe extern "C++" {
+        #[doc(hidden)]
+        #[rust_name = "qmessagelogcontext_default"]
+        fn construct() -> QMessageLogContext;
+    }
+}
+
+#[repr(C)]
+pub struct QMessageLogContext {
+    version: i32,
+    line: i32,
+    file: *const c_char,
+    function: *const c_char,
+    category: *const c_char
+}
+
+impl Default for QMessageLogContext {
+    fn default() -> Self {
+        ffi::qmessagelogcontext_default()
+    }
+}
+
+impl QMessageLogContext {
+    pub fn line(&self) -> i32 {
+        ffi::line(&self)
+    }
+
+    pub fn set_line(&mut self, line: i32) {
+        ffi::set_line(self, line);
+    }
+
+    pub fn file(&self) -> &CStr {
+        unsafe {
+            CStr::from_ptr(ffi::file(self))
+        }
+    }
+
+    pub fn set_file(&mut self, file: &CStr) {
+        unsafe {
+            ffi::set_file(self, file.as_ptr());
+        }
+    }
+
+    pub fn function(&self) -> &CStr {
+        unsafe {
+            CStr::from_ptr(ffi::function(self))
+        }
+    }
+
+    pub fn set_function(&mut self, function: &CStr) {
+        unsafe {
+            ffi::set_function(self, function.as_ptr());
+        }
+    }
+
+    pub fn category(&self) -> &CStr {
+        unsafe {
+            CStr::from_ptr(ffi::category(self))
+        }
+    }
+
+    pub fn set_category(&mut self, category: &CStr) {
+        unsafe {
+            ffi::set_category(self, category.as_ptr());
+        }
+    }
```

Comment:
I'm not really happy with the string handling here, but I think this might be the best we can do. The string must live long enough for it to be used later by the message handler, so it can't be a temporary object (or else the memory is pointing to nothing.)

Might need to add some documentation clarifying this?

Reply:
If you take `&str` here as the type you can use that in CXX, then in C++ you could use `rust::Str::data()` https://cxx.rs/binding/str.html to get the `const char*`

Reply:
Or if it does need to live long enough can it be a `String` and take ownership of it somehow? then use rust::String::data() in C++

Reply:
Actually `&str` might work quite well, as with the lifetimes specified to CXX as well (`fn set_category<'a>(self: &'a mut QMessageLogContext, category: &'a str)` ) to bind it to QMessageLogContext it would force you to do the following or use a static str.
```rust
let string = ...;
context.set_category(&string);
```

wonder if the lifetimes would be needed or if it'd work without, something to experiment with.

Reply:
Managed to work through this, now it's working as expected and won't allow you to create temporaries.